src/python/pants/engine/internals/build_files.py

```diff
@@ -387,16 +389,91 @@ def _extract_env_vars(
     )
 
     # Process synthetic targets.
-    for address_map in address_maps:
-        for synthetic in synthetic_address_maps:
-            synthetic.process_declared_targets(address_map)
-            synthetic.apply_defaults(frozen_defaults)
+
+    def apply_defaults(tgt: TargetAdaptor) -> TargetAdaptor:
+        default_values = frozen_defaults.get(tgt.type_alias)
+        if default_values is None:
+            return tgt
+        return tgt.with_new_kwargs(**default_values, **tgt.kwargs)
+
+    name_to_path_and_synthetic_target: dict[str, tuple[str, TargetAdaptor]] = {}
+    for synthetic_address_map in synthetic_address_maps:
+        for name, target in synthetic_address_map.name_to_target_adaptor.items():
+            name_to_path_and_synthetic_target[name] = (
+                synthetic_address_map.path,
+                apply_defaults(target),
+            )
+
+    name_to_path_and_declared_target: dict[str, tuple[str, TargetAdaptor]] = {}
+    for declared_address_map in declared_address_maps:
+        for name, target in declared_address_map.name_to_target_adaptor.items():
+            name_to_path_and_declared_target[name] = (declared_address_map.path, target)
+
+    # We copy the dict so we can modify the original in the loop.
+    for name, (
+        declared_target_path,
+        declared_target,
+    ) in name_to_path_and_declared_target.copy().items():
+        declared_target_kwargs = dict(declared_target.kwargs)
+        extend_synthetic = declared_target_kwargs.pop("_extend_synthetic", False)
+        if extend_synthetic:
+            # Pop synthetic target to let the declared target take precedence.
+            synthetic_target_path, synthetic_target = name_to_path_and_synthetic_target.pop(
+                name, (None, None)
+            )
```

Comment:
This should be done regardless of `extend_synthetic`, otherwise an explicit target in a BUILD file can't override/replace a synthetic target.

Reply:
Compare with the removed code 
https://github.com/pantsbuild/pants/pull/21725/files#diff-7fd2d0168092a523977f430b4407e923f3b673c15ec2053e4db75c247c89c177L170-L176

where the pop is done before the extend check.

Reply:
Ah good catch!